### PR TITLE
Support for target with class probs in CrossEntropyLoss

### LINKED ISSUE #11959: [feature request] Support soft target distribution in cross entropy loss
Currently our cross entropy loss (i.e., [nn.CrossEntropyLoss](https://pytorch.org/docs/stable/nn.html#crossentropyloss)) only supports a hard target class, i.e., wanting to maximize the output (log) probability of a particular class. But in many times training w.r.t. a soft target distribution (i.e., wanting the output to match a particular distribution) is quite useful too, e.g., preventing overfitting.

## Math
Cross entropy loss operates on logits after softmax. 

Denote the input vector as `x`. Log softmax computes a vector `y` of same length as `x`, where `y_i = x_i - log( \sum_j exp(x_j) )`, representing the log likelihood of each class.

+ In the hard target case, if the target clss is `c`, the loss is simply negative log likelihood loss `-y_c`.

+ In the soft target case, let the target distribution vector be `p` (i.e., `p_i` is the target probability for predicting class `i`). The loss is the KL divergence 

  ```
  D( softmax(x) || p) = \sum_i p_i (log p_i  / softmax(x)_i) = -\sum_i p_i y_i + constant
  ```
  
  The constant is independent of `x` and thus discarded. Our loss formula is just `-\sum_i p_i y_i`. 
  
  When `p_c = 1` for some class `c`, this simplifies to the hard target class.
  

  The formula for gradient computation can be easily derived from this:
  
  ```
  
  d l / d y_i = -p_i
  
  d y_i / d x_i = 1 - exp(x_i) / \sum_j exp(x_j) = 1 - exp(y_i)
  
  # suppose k != i
  d y_k / d_x_i = -exp(x_i) / \sum_j exp(x_j) = - exp(y_i)
  
  # so
  d l / d x_i = exp(y_i) (\sum p) - p_i = exp (y_i) - p_i (= softmax(x) - p_i).
  
  ```

## Possible Implementation

Currently our cross entropy loss implementation takes in batched `x` of shape `(N, C)` and floating point dtype (`N` is the batch size and `C` is the number of classes), and a batched target class indices vector `target` of shape `(N)`, where `target[i]` is the index of the desired output class, and dtype `long` (an integral type).

Since we want it to also take in soft target distribution as target, we can allow it to also take in `target` as a target batched distribution of shape `(N, C)`, and detect whether we want soft target or hard target basing on shape and dtype.

cc @gchanan

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #11959

Alternative approach to creating a new `CrossEntropyLossWithSoftLabels` class. This PR simply adds support for "soft targets" AKA class probabilities to the existing `CrossEntropyLoss` and `NLLLoss` classes.

Implementation is dumb and simple right now, but future work can add higher performance kernels for this case.

### CODE REVIEW

File: torch/nn/modules/loss.py [hunk 1]
```diff
@@ -108,40 +108,61 @@ class NLLLoss(_WeightedLoss):
     The `input` given through a forward call is expected to contain
     log-probabilities of each class. `input` has to be a Tensor of size either
     :math:`(minibatch, C)` or :math:`(minibatch, C, d_1, d_2, ..., d_K)`
-    with :math:`K \geq 1` for the `K`-dimensional case (described later).
+    with :math:`K \geq 1` for the `K`-dimensional case. The latter is useful for
+    higher dimension inputs, such as computing NLL loss per-pixel for 2D images.
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/modules/loss.py:112]: Great point, and I agree with you. AFAICT it doesn't make mathematical sense to define "soft labels" for `NLLLoss` since likelihood requires a single specific outcome, as you mentioned. I'll update the PR to be `CrossEntropyLoss`-only.

Note that this will make `CrossEntropyLoss` no longer equivalent to `LogSoftmax` + `NLLLoss`; the docs will need to reflect that they are only equivalent for the case of class index targets.

File: torch/nn/modules/loss.py [hunk 2]
```diff
@@ -108,40 +108,61 @@ class NLLLoss(_WeightedLoss):
...
+
+      .. math::
+          \ell(x, y) = \begin{cases}
+              \sum_{n=1}^N \frac{1}{\sum_{n=1}^N w_{y_n} \cdot \mathbb{1}\{y_n \not= \text{ignore\_index}\}} l_n, &
+               \text{if reduction} = \text{`mean';}\\
+                \sum_{n=1}^N l_n,  &
+                \text{if reduction} = \text{`sum'.}
+            \end{cases}
+
+    - Probabilities over classes; useful when labels beyond a single class per minibatch item
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/modules/loss.py:142]: Hm I agree with making it clear that there's one probability per class, but "vector" is a bit confusing to me because the target will be N-dimensional (same as the input shape).

What do you think about just "Probabilities for each class" instead?

File: aten/src/ATen/native/LossNLL.cpp [hunk 3]
```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
...
+      !weight.defined() || (weight.dim() == 1 && weight.numel() == n_classes),
+      "weight tensor should be defined either for all ",
+      n_classes,
+      " classes or no classes"
+      " but got weight tensor of shape: ",
+      weight.sizes());
+
+  Tensor ret;
+  Tensor input = at::log_softmax(
+      self, 1, optTypeMetaToScalarType(self.options().dtype_opt()));
```
Comment (zou3519) [inline aten/src/ATen/native/LossNLL.cpp:474]: optTypeMetaToScalarType(self.options().dtype_opt()) -> `self.scalar_type` should work

File: aten/src/ATen/native/LossNLL.cpp [hunk 4]
```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
...
 
+Tensor cross_entropy_loss_prob_target(
+    const Tensor& self,
+    const Tensor& target,
+    const Tensor& weight,
+    int64_t reduction) {
+  const auto n_classes = self.size(1);
+  TORCH_CHECK(
+      !weight.defined() || (weight.dim() == 1 && weight.numel() == n_classes),
+      "weight tensor should be defined either for all ",
```
Comment (zou3519) [inline aten/src/ATen/native/LossNLL.cpp:466]: nit: say the name of the operation: "cross_entropy: weight tensor should be..."

File: aten/src/ATen/native/LossNLL.cpp [hunk 5]
```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
...
+      n_classes,
+      " classes or no classes"
+      " but got weight tensor of shape: ",
+      weight.sizes());
+
+  Tensor ret;
+  Tensor input = at::log_softmax(
```
Comment (zou3519) [inline aten/src/ATen/native/LossNLL.cpp:473]: nit: auto input

File: aten/src/ATen/native/LossNLL.cpp [hunk 6]
```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
...
+    } else if (reduction == Reduction::Sum) {
+      ret = ret.sum();
+    }
+  } else {
+    ret = -(input * target).sum(1);
+    if (reduction == Reduction::Mean) {
+      ret = ret.mean();
+    } else if (reduction == Reduction::Sum) {
+      ret = ret.sum();
+    }
```
Comment (zou3519) [inline aten/src/ATen/native/LossNLL.cpp:494]: We can make this faster by making the cases separate:
```
switch (reduction) {
  case Reduction::Mean:
    ret = -(input * target).sum(1)
  case Reduction::Sum:
    ret = -(input * target).sum()
  case Reduction::None:
    ret = -(input * target).mean()
  default:
    // assert false
}

```

File: aten/src/ATen/native/LossNLL.cpp [hunk 7]
```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
...
+  if (weight.defined()) {
+    // Expand weight to the correct number of dims for broadcasting with input / target
+    auto weight_bc_shape = std::vector<int64_t>(input.dim(), 1);
+    weight_bc_shape[1] = weight.size(0);
+    Tensor weight_ = weight.view(weight_bc_shape);
+
+    ret = -(input * target * weight_).sum(1);
+    if (reduction == Reduction::Mean) {
+      // Compute weighted mean
+      ret = ret.sum() / (target * weight_).sum();
```
Comment (zou3519) [inline aten/src/ATen/native/LossNLL.cpp:484]: This... seems like a good argument for why a user would expect reduction=mean to return (-(input * target * weight_).mean()), I'm having a hard time coming up with a use case where someone wants probabilities and wants to do a weighted mean over the probabilities and weights.

At any rate we should probably be consistent with our hard targets cross_entropy function...

File: aten/src/ATen/native/LossNLL.cpp [hunk 8]
```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
...
+    } else if (reduction == Reduction::Sum) {
+      ret = ret.sum();
+    }
```
Comment (zou3519) [inline aten/src/ATen/native/LossNLL.cpp:487]: We can make this faster by making all the cases separate; no need to reduce partially and then reduce some more

File: aten/src/ATen/native/LossNLL.cpp [hunk 9]
```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
...
+      " classes or no classes"
+      " but got weight tensor of shape: ",
+      weight.sizes());
+
+  Tensor ret;
+  Tensor input = at::log_softmax(
+      self, 1, optTypeMetaToScalarType(self.options().dtype_opt()));
+  if (weight.defined()) {
+    // Expand weight to the correct number of dims for broadcasting with input / target
+    auto weight_bc_shape = std::vector<int64_t>(input.dim(), 1);
```
Comment (zou3519) [inline aten/src/ATen/native/LossNLL.cpp:477]: nit: use SmallBuffer instead of std::vector to avoid dynamic allocations: https://github.com/pytorch/pytorch/blob/master/c10/util/SmallBuffer.h

File: aten/src/ATen/native/LossNLL.cpp [hunk 10]
```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
...
     int64_t reduction,
     int64_t ignore_index) {
-  return at::nll_loss_nd(
-      at::log_softmax(
-          self, 1, optTypeMetaToScalarType(self.options().dtype_opt())),
-      target,
-      weight,
-      reduction,
-      ignore_index);
+  Tensor ret;
```
Comment (zou3519) [inline aten/src/ATen/native/LossNLL.cpp:505]: nit: you can just return inside each of the branches of the conditional to save two lines of code

File: torch/testing/_internal/common_nn.py [hunk 11]
```diff
@@ -4695,6 +4720,36 @@ def padding3d_circular(input, pad):
...
+    dict(
+        module_name='CrossEntropyLoss',
+        input_size=(5, 3, 4, 2),
+        target_fn=lambda: torch.rand(5, 3, 4, 2).softmax(dim=1),
+        reference_fn=lambda i, t, m:
+            loss_reference_fns['CrossEntropyLoss'](i, t, reduction=get_reduction(m)),
+        check_sum_reduction=True,
+        desc='4d_prob_target',
+        check_bfloat16=False,
+    ),
```
Comment (zou3519) [inline torch/testing/_internal/common_nn.py:4752]: Do any of these test reduction=Mean?

File: torch/testing/_internal/common_nn.py [hunk 12]
```diff
@@ -4695,6 +4720,36 @@ def padding3d_circular(input, pad):
...
+    dict(
+        module_name='CrossEntropyLoss',
+        input_size=(5, 3),
+        target_fn=lambda: torch.rand(5, 3).softmax(dim=1),
+        reference_fn=lambda i, t, m:
+            loss_reference_fns['CrossEntropyLoss'](i, t, reduction=get_reduction(m)),
+        check_sum_reduction=True,
+        desc='2d_prob_target',
+        check_bfloat16=False,
+    ),
```
Comment (zou3519) [inline torch/testing/_internal/common_nn.py:4732]: We should test that `weights` work as expected with a probability target; I don't see weights being used by these tests

File: torch/nn/modules/loss.py [hunk 13]
```diff
@@ -138,11 +139,6 @@ class NLLLoss(_WeightedLoss):
             \text{if reduction} = \text{`sum'.}
         \end{cases}
 
-    Can also be used for higher dimension inputs, such as 2D images, by providing
-    an input of size :math:`(minibatch, C, d_1, d_2, ..., d_K)` with :math:`K \geq 1`,
-    where :math:`K` is the number of dimensions, and a target of appropriate shape
-    (see below). In the case of images, it computes NLL loss per-pixel.
-
```
Comment (zou3519) [inline torch/nn/modules/loss.py:145]: Is this deletion intentional?

File: torch/nn/functional.py [hunk 14]
```diff
@@ -2790,7 +2791,9 @@ def cross_entropy(
             when reduce is ``False``. Default: ``True``
         ignore_index (int, optional): Specifies a target value that is ignored
             and does not contribute to the input gradient. When :attr:`size_average` is
-            ``True``, the loss is averaged over non-ignored targets. Default: -100
+            ``True``, the loss is averaged over non-ignored targets. Note that
+            :attr:`ignore_index` is only applicable when the target contains class indices.
+            Default: -100
```
Comment (zou3519) [inline torch/nn/functional.py:2796]: We should also update the example in the functional version (on line 2810 of this file)

File: torch/nn/functional.py [hunk 15]
```diff
@@ -2778,9 +2778,10 @@ def cross_entropy(
             in case of 2D Loss, or :math:`(N, C, d_1, d_2, ..., d_K)` where :math:`K \geq 1`
             in the case of K-dimensional loss. `input` is expected to contain unnormalized scores
             (often referred to as logits).
-        target (Tensor) : :math:`(N)` where each value is :math:`0 \leq \text{targets}[i] \leq C-1`,
-            or :math:`(N, d_1, d_2, ..., d_K)` where :math:`K \geq 1` for
-            K-dimensional loss.
+        target (Tensor) : If containing class indices, shape :math:`(N)` where each value is
+            :math:`0 \leq \text{targets}[i] \leq C-1`, or :math:`(N, d_1, d_2, ..., d_K)` with
+            :math:`K \geq 1` in the case of K-dimensional loss. If containing class probabilities,
+            same shape as the input.
```
Comment (zou3519) [inline torch/nn/functional.py:2784]: F.cross_entropy docs say "This criterion combines `log_softmax` and `nll_loss` in a single function" which is no longer true

**Review by zou3519 — COMMENTED** [review]
this looks pretty good, some suggestions and comments

File: torch/nn/modules/loss.py (hunk 13 above)
Comment (jbschlosser) [reply torch/nn/modules/loss.py:145]: Yes, I updated the `NLLLoss` docs slightly to be consistent with `CrossEntropyLoss` for this part. This part is mostly a repeat of the description of K-dimensional loss further up.

File: torch/testing/_internal/common_nn.py (hunk 11 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_nn.py:4752]: Nope, unfortunately this machinery doesn't support `reduction='mean'`. I added a test in `test_nn.py` that tests all reduction types, but it's only for the one-hot case. I'll add another set for the non-one-hot case.

File: torch/testing/_internal/common_nn.py (hunk 12 above)
Comment (jbschlosser) [reply torch/testing/_internal/common_nn.py:4732]: The one-hot test in `test_nn.py` uses weights but I'll also add them here.

File: aten/src/ATen/native/LossNLL.cpp (hunk 7 above)
Comment (jbschlosser) [reply aten/src/ATen/native/LossNLL.cpp:484]: Yeah, I agree it's super weird, but I did it this way to be consistent with the hard target cross-entropy loss. With one-hot targets, the results are equivalent between soft and hard if done like this :/

File: aten/src/ATen/native/LossNLL.cpp (hunk 6 above)
Comment (jbschlosser) [reply aten/src/ATen/native/LossNLL.cpp:494]: Split the cases, but note that `ret = -(input * target).sum(1).mean()` is required to properly compute the mean case.

File: aten/src/ATen/native/LossNLL.cpp (hunk 10 above)
Comment (jbschlosser) [reply aten/src/ATen/native/LossNLL.cpp:517]: does this break NRVO though?

**Review by zou3519 — APPROVED** [review]
Some minor comments but otherwise this LGTM!

File: torch/nn/modules/loss.py [hunk 16]
```diff
@@ -1019,46 +1014,70 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
 
 
 class CrossEntropyLoss(_WeightedLoss):
-    r"""This criterion combines :class:`~torch.nn.LogSoftmax` and :class:`~torch.nn.NLLLoss` in one single class.
+    r"""This criterion computes the cross entropy loss between input and target.
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/modules/loss.py:1017]: Sure enough, the C++ module still mentioned being a combination of LogSoftmax + NLLLoss. Fixed

File: torch/nn/modules/loss.py [hunk 17]
```diff
@@ -1019,46 +1014,70 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
...
+      :math:`N` is the batch size, and :math:`C` is the number of classes. If
+      :attr:`reduction` is not ``'none'`` (default ``'mean'``), then
+
+      .. math::
+          \ell(x, y) = \begin{cases}
+              \frac{\sum_{n=1}^N l_n}{N}, &
+               \text{if reduction} = \text{`mean';}\\
+                \sum_{n=1}^N l_n,  &
+                \text{if reduction} = \text{`sum'.}
+            \end{cases}
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply torch/nn/modules/loss.py:1075]: Yeah, `N` is doing a lot of work implicitly here. I do think that data of shape `(N, C, d1, d2, ..., dk)` is conceptually a batch of `(N, d1, d2, ..., dk)` distributions (and as mentioned before, I think `d1, ..., dk` should have been added to the left of `C` for this to be clearer, but that ship has sailed).

While this was carried over to some extent from the old docs, each item in the formula is now more explicitly defined, so I think it needs to be more precise. Specifically, "N is the batch size" should change. Borrowing some terminology from `KLDivLoss`, we could do something like:

```
N spans the minibatch dimension as well as dimensions d1, ..., dk in the case of K-dimensional loss
```

wdyt?

[the remark this answers is not in the document]
Comment (zou3519) [reply torch/nn/modules/loss.py:1075]: > N spans the minibatch dimension as well as dimensions d1, ..., dk in the case of K-dimensional loss

That sounds good

**Comment by VitamintK** [comment]
the commit message in https://github.com/pytorch/pytorch/commit/a42345adee3eb32a6be3eb6863cac886ed8f795f (from the first post of this PR) says `This PR simply adds support for "soft targets" AKA class probabilities to the existing `CrossEntropyLoss` and `NLLLoss` classes.`, but it actually only adds the change for `CrossEntropyLoss`, right?

**Comment by jbschlosser** [comment]
> the commit message in [a42345a](https://github.com/pytorch/pytorch/commit/a42345adee3eb32a6be3eb6863cac886ed8f795f) (from the first post of this PR) says `This PR simply adds support for "soft targets" AKA class probabilities to the existing `CrossEntropyLoss`and`NLLLoss` classes.`, but it actually only adds the change for `CrossEntropyLoss`, right?

That's right; it was only added to `CrossEntropyLoss`.
